### ISSUE TITLE
[bug] conan2 self-loop when using tool_requires in profiles

### ISSUE THREAD

**Issue by thorsten-klein** [conan-io/conan#10219]
### Environment Details (include every applicable attribute)
  * Operating System+version: Linux 20.04
  * Compiler+version: gcc 9.3
  * Conan version: develop2
  * Python version: 3.8.10

### Steps to reproduce (Include if Applicable)
testprofile.txt
```
[tool_requires]
tool1/0.1@test/testchannel
tool2/0.1@test/testchannel
tool3/0.1@test/testchannel
```

this results in a self-loop if applied for build and host context both:

### Logs (Executed commands with output) (Include/Attach if Applicable)
`conan info . --graph={graphfile} -pr:b=default -pr:h=default -pr:b=testprofile.txt -pr:h=testprofile.txt`
This command will never finish because of a self-loop.  

The graph currently would look like:
```
digraph {
        "conanfile.py" -> "tool1/0.1@test/testchannel"
        "conanfile.py" -> "tool2/0.1@test/testchannel"
        "conanfile.py" -> "tool3/0.1@test/testchannel"
        "tool2/0.1@test/testchannel" -> "tool1/0.1@test/testchannel"
        "tool3/0.1@test/testchannel" -> "tool1/0.1@test/testchannel"
        "tool1/0.1@test/testchannel" -> "tool2/0.1@test/testchannel"
        "tool3/0.1@test/testchannel" -> "tool2/0.1@test/testchannel"
        "tool1/0.1@test/testchannel" -> "tool3/0.1@test/testchannel"
        "tool2/0.1@test/testchannel" -> "tool3/0.1@test/testchannel"
    ..........
}
```

We need to break this loop, e.g. `tool1 -> tool2 -> tool1 -> ...`

### Proposal
We should introduce some pattern `&+` (analog to `&` or `&!`) which means `applied for all consumers + its build_requirements, but not for other tool_requires from profiles`

newtestprofile.txt
```
tool1/0.1@test/testchannel      <-- will be applied for everything
&+: tool2/0.1@test/testchannel <-- will only be applied for the consumer and its build_requires, but not for tool1 or tool3
&: tool3/0.1@test/testchannel <-- will only be applied for the consumer
```

`conan info . --graph={graphfile} -pr:b=default -pr:h=default -pr:b=newtestprofile.txt -pr:h=newtestprofile.txt`
will result in

```
digraph {
        "conanfile.py" -> "tool1/0.1@test/testchannel"
        "conanfile.py" -> "tool2/0.1@test/testchannel"
        "conanfile.py" -> "tool3/0.1@test/testchannel"
        "tool2/0.1@test/testchannel" -> "tool1/0.1@test/testchannel"
        "tool3/0.1@test/testchannel" -> "tool1/0.1@test/testchannel"
}
```

The next step would be to exclude also some specific packages from the pattern, e.g.
newertestprofile.txt
```
*: tool1/0.1@test/testchannel      <-- will be applied for everything
*,!tool1/*: tool2/0.1@test/testchannel <-- will be applied for everything, but not tool1
&+: tool3/0.1@test/testchannel <-- will only be applied for the consumer but not for tool1 or tool2
```

PS: `&+` is actually the default behavior of conan 1.x --> The build_requires from profiles are not applied to build_requires from profiles. 
Maybe we can discuss if we want to make this bahavior default in conan 2.0 also, since otherwise many people will face the loop issue. As a result people MUST consider the `&+` pattern explicitly to break the loop. 
Otherwise people MUST only explicitlty use some pattern (e.g. `*+`) if they want to apply the tool_requires for really everything - so they cause the loop explicitly.

**Comment by memsharded** [comment]
We have been discussing this, and a some other related things have changed.

First, the most important point is that using the 2 profiles approach, it doesn't make sense to have the same ``tool_requires`` defined both in the host and the build profile. For example ``cmake/3.19`` is a tool require that you want to apply to the "host" context, but not necessarily to other tools in the "build" context. The decoupling between the host and build context is the first thing to leverage.

Then, in the build context, we will typically find very few packages, for example ``cmake/3.19``, ``meson/0.60``, etc. If some of them require in turn other ``tool_requires`` it is possible to define now them in two ways:

- Specify them directly in the recipe. This might be a good option, specially if you own the packages.
- Specify them in the "build" profile, and positively adding the ones needed. Lets say that ``cmake/3.19`` needs ``cmake/3.18`` to build (this is not possible in 1.X but might be possible in 2.0), then this might be even possible (not saying desirable, but possible):
  ```
  [tool_requires]
  cmake/3.19: cmake/3.18
  ```

In case that the positive patterns are not enough, negative patterns have also been added in Conan 2.0, so something like:
```
# this is the build profile
[tool_requires]
!cmake*: othertool/1.0
```
will also be possible (meaning that it will inject the ``othertool/1.0`` to every package in the build context except the ``cmake`` ones.

With these features, it seems that managing the ``tool_requires`` in 2.0 will be reasonable enough. Please have a look, and if you have other use case, it would be useful to learn it with real package names instead of ``tool1/1.0``, but ``cmake/3.19``, to understand that use case better. Thanks!

**Comment by memsharded** [comment]
As commented above: 
> With these features, it seems that managing the tool_requires in 2.0 will be reasonable enough. Please have a look, and if you have other use case

So closing this ticket as addressed by 2.0, thanks!
